rfcs/026-work-enricher/README.md

```diff
@@ -0,0 +1,160 @@
+
+* The matcher currently uses a custom DynamoDB graph implementation which was built specifically for the purpose. This was never meant to be a general solution and a rewrite would likely need to be involved in the building of some common system.
+
+* The graph element of both the matcher and relations work are relatively similar, but the denormalisation stages (the enricher and the merger) have less parallels. For example, the input to the merger is currently a group of work IDs to be merged with the stage having no knowledge of the graph representation, unlike the enricher which is likely to receive single works and needs access to the graph.
+
+Besides from the use of graphs and similarities to the matcher and merger, another aspect worth mentioning is the fact that we will need another store for denormalised works in addition to a store for merged works. Having two Elasticsearch indices coupled to corresponding ingestors is a possible solution (one for the current index of merged works and adding an additional one for denormalised works). There have been some previous discussions about replacing the recorder VHS with an Elasticsearch index, so it might be a worthwhile time to consider how this would work in this context, as here we would introducing an Elasticsearch index which was not the sink of the pipeline but used as an intermediate store.
+
+## Graph Store
+
+There has been some investigation into the use of graph databases such as Neo4j and AWS Neptune for storage and analysis of concepts in the pipeline. Using a specialised graph store might be a good general solution for relationships within our data.
+
+Neo4j for example would enable us, if desired, to store heterogeneous data types and relationships within a single graph:
+
+```
+[ Work A ] ── (matches with) ── [ Work C ]
+    │
+(parent of)
+    │
+[ Work B ]
+    │
+(relates to)
+    │
+    ├── [ Concept A ]
+    └── [ Concept B ]
+```
+
+Here square brackets indicate entities and parentheses indicate (directed) relationships.
+
+Having a single graph representation would enable richer queries and analytics of our data, and would likely be especially useful for machine learning work. However this may introduce complexity by breaking assumptions about data flow in the pipeline, such as if there were multiple stages which had the ability to write to the graph. This should therefore should be considered with caution. Whether as part of a single graph representation or a more specialised store solely for inter work relationships in the context of archives, using such a database to store relationships would aid implementation of the denormalisation stage, especially with the traversal and emitting of related works.
```

Comment:
I'm interested in how much can be done with pure graph DB queries. The `query` -> `process results` -> `update graph` flow is so natural for graph DBs that I _think_ most of them can do this kind of work within the query language, which obviously has massive advantages not least in terms of consistency.

[This](https://neo4j.com/docs/cypher-manual/current/introduction/quering-updating-administering/) is one thing I could find, I'd be interested to see if we could write a pseudo-query doing the sort of thing we're going to want.

Reply:
The enricher stage would only have read access to the graph store, with messages at its output being the way it affects the outside world, so don't think these sort of queries are going to be useful.

Reply:
But if we have a graph store we will have something that writes to it - which will need to do something like this (like the matcher does now), no?

Reply:
Yes true. It is just adding single connections and does not need to do full graph traversal, but something like that could still be useful.